### PR TITLE
Improve AI heroes interaction with castles and heroes in dangerous areas

### LINKED ISSUE #7082: Adventure map, threatened AI is running between two points
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

Threatened AI is trying to get to some point, but the next turn returns back to castle.

https://user-images.githubusercontent.com/55348946/235298669-42cfb856-91a5-4887-8d5a-837130d1de22.mp4

If strong hero runs away everything becomes OK.

https://user-images.githubusercontent.com/55348946/235298723-4c24a5db-670f-43a8-a3f3-16be18a47ab2.mp4

### Save file

[Scared AI.zip](https://github.com/ihhub/fheroes2/files/11358958/Scared.AI.zip)

### Additional info

_No response_

### LINKED ISSUE #7193: Adventure map, AI behavior, AI does not protect its castles enough (1.0.4 7171)
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

In fheroes2, AI does not protect its castles enough.

I don't know if it's because of a lack of gold and/or resources, but the castle's garrison is not very numerous, while there are a significant number of creatures available for recruitment.

Also, in the example, the halflings should be spread across all the slots rather than just one to maximize the damage possible with the ballista and turrets before all the halflings groups are decimated.

Also, the main hero goes away from the castle to visit the resource generators, while this could be done by a secondary hero.

My heroes may seem far away, but at least one of them, Sandro, can easily reach the AI player's castle in one turn with the Dimension Door spell.

In fheroes2:

AI castle poor defense 001

AI castle poor defense 002

AI castle poor defense 003

AI castle poor defense 004

### Describe the possible solution

In this case, the powerful hero of the AI should be positioned in defense of the castle with the maximum number of recruited creatures, as he does not have many other possibilities of action from a strategic point of view on the adventure map being stuck on his island.
A secondary hero should take care of collecting resources from the resource generators.
Also, in this case, there is no point in saving gold, the castle is fully built.
AI might as well spend the gold on buying creatures waiting for the enemy to attack or on building up a large army while waiting for other possible strategic actions, such as capturing the boat of a weak enemy hero who ventures onto the island, or recruiting a hero who has the Dimension Door spell.
AI should also take into account the possible use of "long distance" movement spells (notably dimension door) by its opponents to organize the defense of its castles and not wait for an opposing hero to be a threat near the castle.

### Additional info

Here are the save files (end of previous turn, beginning of turn, and in front of the AI player's castle):
[AI castle poor defense 01.zip](https://github.com/ihhub/fheroes2/files/11497460/AI.castle.poor.defense.01.zip)

### PR DESCRIPTION
**Pull request by ihhub**
This pull request fixes multiple things in AI behavior:
- move to a castle and stay there the hero cannot defeat the treat (close #7082). This is important to stay in the castle if the hero is too weak
- capture empty castles even if the area is under treat (close #7193). Also be afraid only of heroes which are really more powerful that the hero

### CODE REVIEW

**Review by oleg-derevenetz — APPROVED** [review]

**Comment by ihhub** [comment]
I'm marking this pull request as draft as with the all changes on some maps heroes are too defensive and never leave their castles even if heroes are very powerful.

File: src/fheroes2/ai/normal/ai_normal_kingdom.cpp [hunk 1]
```diff
@@ -555,10 +646,15 @@ namespace AI
                     if ( !castle )
                         continue;
 
-                    const Army & castleArmy = castle->GetArmy();
-                    enemyArmies.emplace_back( idx, &castleArmy );
+                    if ( !castle->isCastle() ) {
+                        // If it is just a town then there is no way to hire heroes which can be a threat.
```
Comment (oleg-derevenetz) [inline src/fheroes2/ai/normal/ai_normal_kingdom.cpp:650]: But enemy can upgrade it to a castle and hire a hero during the same turn, right?

Comment (ihhub) [reply src/fheroes2/ai/normal/ai_normal_kingdom.cpp:650]: Good point. I think we need to add an extra condition if a castle cannot be built.

Comment (ihhub) [reply src/fheroes2/ai/normal/ai_normal_kingdom.cpp]: It is fixed.

**Comment by ihhub** [comment]
Hi @Branikolog , please review this pull request as the highest priority.

**Review by Branikolog — APPROVED** [review]
👍

**Pull request merged**